CLDC-2290 implement delete multiple logs story

arfacamble opened the pull request:
[ticket](https://digital.dclg.gov.uk/jira/browse/CLDC-2290)

All routes and methods for this story have been extracted to own controller.

|<img width="1040" alt="image" src="https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/assets/51094020/e34f908c-8897-45f8-a979-1b3c3e5e7860">
|-
|-
|<img width="1117" alt="image" src="https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/assets/51094020/cf5c5907-7427-481b-8197-610a0d3e3f8e">
|-
|<img width="800" alt="image" src="https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/assets/51094020/0f1e9089-c097-439f-a852-1339c521060e">
|-
|<img width="1206" alt="image" src="https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/assets/51094020/fad8cd2e-b2c4-4f8e-aff9-6f691b31179e">
|-

arfacamble committed:
add a button to the logs list to delete multiple logs
style and position of button
helpers for displaying the button conditionally depending on user role and what filters and search are active

arfacamble committed:
correct indentation from 4 spaces to 2 in view file

arfacamble committed:
test appearance of delete logs button on index page for lettings logs

arfacamble committed:
write a happy path feature test for the entire journey

jamesrose commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -111,6 +110,37 @@ def update_logs
+    @delete_logs_form = Forms::DeleteLogsForm.new(log_ids:, log_type:)
+    @search_term = params.dig(:forms_delete_logs_form, :search_term)
+    if @delete_logs_form.no_logs_selected?
+      redirect_to delete_logs_lettings_logs_path(search: @search_term, no_logs_selected: true)
```
We shouldn't need to redirect here. Instead, we can just re-render the template with errors.

jamesrose commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -111,6 +110,37 @@ def update_logs
+    @search_term = search_term
+    @delete_logs_form = Forms::DeleteLogsForm.new(logs: filtered_logs(all_logs, @search_term, @session_filters))
+    if params[:no_logs_selected] == "true"
+      @delete_logs_form.errors.add(:logs_ids, "Select at least one log to delete or press cancel to return")
```
This logic belongs in the form object itself.

jamesrose commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -111,6 +110,37 @@ def update_logs
+  def delete_logs_confirmation
+    log_ids = params.dig(:forms_delete_logs_form, :logs_to_delete)
+    log_type = params.dig(:forms_delete_logs_form, :log_type).to_sym
+    @delete_logs_form = Forms::DeleteLogsForm.new(log_ids:, log_type:)
```
```suggestion
    @delete_logs_form = Forms::DeleteLogsForm.new(attributes: params[:forms_delete_logs_form])
```

The extraction of these attributes should be encapsulated within the form object.

jamesrose commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -111,6 +110,37 @@ def update_logs
+      log = LettingsLog.visible.find(id)
+      authorize log, :destroy?
+      log.discard!
+    end
```
I'm curious why we need to split IDs here?

```suggestion
  def delete_logs
    logs = LettingsLog.visible.find(params[:log_ids])
    logs.each { |log| authorize(log, :destroy?) }
    logs.discard_all!
  end
```

jamesrose commented on app/models/forms/delete_logs_form.rb:
```diff
@@ -0,0 +1,18 @@
+
+    def no_logs_selected?
+      @log_ids.blank? || @log_ids.reject(&:blank?).blank?
+    end
```
We should add validations here and instead use `valid?` externally to determine whether we're save to progress the deletion.

jamesrose commented on app/models/forms/delete_logs_form.rb:
```diff
@@ -0,0 +1,18 @@
+    def initialize(logs: nil, log_ids: nil, log_type: nil)
+      @logs = logs
+      @log_ids = log_ids
+      @log_type = log_type || (logs.first&.lettings? ? :lettings : :sales)
```
I think this optimisation has introduced complexity. Did you consider two separate form objects and templates for now?

arfacamble replied:
Given that in the confirmation method, we need to create a form object to ask if it is valid and then either pass that thing back to the list of logs to delete view or on to the confirmation page, I have stuck with one class of form object but admittedly it does have some complexity in its initialize method still. Happy to discuss how that might be mitigated.

jamesrose commented on app/helpers/log_list_helper.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def delete_logs_path(logs, search_term)
+    controller_type = controller.class.name.delete_suffix("Controller").underscore.to_sym
```
We shouldn't have to do this. Can you explain what you're trying to din this method?

arfacamble replied:
I need to insert a Delete Logs button into the index page (_log_list partial) which is reused for all different log types and can be rendered by three different controllers.
I'm up for discussing this and further thoughts on it, but I also think the implementation in anger of this functionality can be worried about when I'm implementing for these other controllers.

asmega replied:
if we want fewer if/case statements if this were moved to the relevant controller or helper the switching can be removed. ie `LettingsLogsController` or `SalesLogsController` (or the helpers, not sure on those names or ordering)

arfacamble replied:
🤦‍♂️

jamesrose commented on app/views/logs/delete_logs.html.erb:
```diff
@@ -0,0 +1,61 @@
+            <% end %>
+            <% row.cell(text: status_tag(log.status)) %>
+            <% row.cell do %>
+              <% f.check_box :bla %>
```
!

jamesrose commented on app/views/logs/delete_logs.html.erb:
```diff
@@ -0,0 +1,61 @@
+            <% row.cell(text: status_tag(log.status)) %>
+            <% row.cell do %>
+              <% f.check_box :bla %>
+              <% f.govuk_check_box :logs_to_delete, log.id,
```
```suggestion
              <% f.govuk_check_box :log_ids, log.id,
```

arfacamble replied:
there is already a variable log_ids representing all the logs in the table.
Do you want the change so that it is expressing that this thing is a collection of ids rather than actual logs?
I think here in the context of this form and consistent with the naming in the controller methods I will change this to selected_ids unless you don't like that

arfacamble committed:
create basic tests for the view component for listing logs to delete

arfacamble committed:
create request tests for the GET delete-logs path

arfacamble committed:
create request tests for the GET delete-logs-confirmation path

arfacamble committed:
create request tests for the DELETE delete-logs path

jamesrose commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -111,6 +111,31 @@ def update_logs
 
+  def delete_logs_list
+    log_ids = params.require(:log_ids).map(&:to_i)
+    selected_ids = params[:selected_ids]&.map(&:to_i)
```
Are you sure that you need to map these to integers? AR finders are quite content with strings.

asmega commented on app/views/logs/delete_lettings_logs.html.erb:
```diff
@@ -0,0 +1,49 @@
+  Review the logs you want to delete
+</h1>
+<% logs = @delete_logs_form.logs %>
+<p>You've selected <%= logs&.count %> <%= "log".pluralize(logs&.count) %> to delete</p>
```
the form can act a view object so i wouldn't bother with the local variable and just reference `@delete_logs_form` wholly instead

as an example if you want less code is to push stuff/logic downwards so the responsibility becomes delegated. if the form implemented `form.log_count` you could push that downwards then view does not need to deal with nil guards. the view just wants to ask for some info ie log count to render. it shouldn't need to deal with nil guards as that is someone else's responsibility, in this case the form/view.object

asmega commented on config/routes.rb:
```diff
@@ -217,6 +222,10 @@
 
+      get "delete-logs", to: "sales_logs#delete_logs_list"
+      get "delete-logs-confirmation", to: "sales_logs#delete_logs_confirmation"
+      delete "delete-logs", to: "sales_logs#delete_logs"
```
how come sales has 3 routes added here but lettings has 4 new routes?

asmega commented on app/models/forms/delete_logs_form.rb:
```diff
@@ -0,0 +1,29 @@
+      all_logs = case @log_type
+                 when :lettings then current_user.lettings_logs.visible
+                 when :sales then current_user.sales_logs.visible
+                 end
```
if you made `current_user` available with a getter ie `attr_reader :current_user` the rest of the code can be simplified. if you're concerned with exposing this then you can make the getter private

if the above is done then `all_logs` can be extracted to a private method giving shorter more self contained and documented methods

arfacamble committed:
comprehensive reworking after code review
ensure that we are not passing lists of ids through params in the query string, risking overflowing the maximum URL length,
adjust tests accordingly,
do not attempt to reuse the same table for sales and lettings

arfacamble committed:
alter config to allow creating controllers from the command line with associated spec files that matches how we test

arfacamble committed:
extract controller methods and associated tests to do with the delete logs feature into their own controller,
    amend routes accordingly

arfacamble committed:
implement same work for sales as for lettings

arfacamble commented on app/views/organisations/logs.html.erb:
```diff
@@ -36,6 +36,7 @@
                  total_count: @total_count,
                  csv_download_url: csv_download_url_by_log_type(@log_type, @organisation, search: @search_term, codes_only: false),
                  csv_codes_only_download_url: csv_download_url_by_log_type(@log_type, @organisation, search: @search_term, codes_only: true),
+                 delete_logs_path: "#",
```
as in slack description, I am now implementing the work you see here for lettings for organisations and sales also. This will be improved

arfacamble commented on spec/requests/lettings_logs_controller_spec.rb:
```diff
@@ -347,7 +347,6 @@
               Timecop.freeze(2022, 3, 1) do
                 example.run
               end
-              Timecop.return
```
very much so.
The point of putting `example.run` in a block is that the time freeze exists only for that block.
The reasonable alternative to what I have changed this to would be:
```
Timecop.freeze(date)
example.run
Timecop.return
```

arfacamble commented on spec/views/form/page_view_spec.rb:
```diff
@@ -22,8 +22,6 @@ def assign_attributes(object, attrs)
       example.run
     end
-    Timecop.return
-    Singleton.__init__(FormHandler)
```
@bibblobcode similarly here Jack, you may not be aware that a while ago @kosiakkatrina put a line in the test config so that the FormHandler singleton is reinitialised after every single test, so anywhere you see this you can also try removing it as it should be redundant

arfacamble committed:
implement the story for lettings and sales logs under the organisation tab
routing and controller methods
testing for deleting sales logs, lettings or sales logs for an organisation
move storage of relevant routes inside the form object as a comprehensive view model

arfacamble commented on config/application.rb:
this allows us to create a controller from the command line and create matching test files for request specs rather than controller tests

arfacamble committed:
merge the delete pages for lettings logs and sales logs, add to the tests for the lettings page to test sales specific content

arfacamble committed:
minor refactor to delete logs controller: ensure session filters are only fetched from teh session when needed and extract discard logs method to private method

jamesrose commented on app/models/forms/delete_logs_form.rb:
```diff
@@ -0,0 +1,40 @@
+      end
+    end
+
+    def all_logs
```
Given this is not _all_ logs, how about renaming to?

```suggestion
    def visible_logs
```

jamesrose commented on app/views/logs/delete_logs.html.erb:
```diff
@@ -0,0 +1,58 @@
+    <% table.head do |head| %>
+      <% head.row do |row| %>
+        <% row.cell(header: true, text: "Log ID") %>
+        <% if @delete_logs_form.log_type == :lettings %>
```
There is enough variation here that you should render a different template for each table, embedded within the form.

arfacamble replied:
I just merged `delete_lettings_logs.html.erb` and `delete_sales_logs.html.erb` yesterday as after I had polished everything then they had two lines different.
I think I will need a little convincing to change this back as as far as I can tell what you're suggesting is still having an if/else inside the form to render two different table partials that only have 2 lines out of ~30 different, so it feels like much of a muchness to me.

jamesrose replied:
You could render the template based on the type, which would avoid the need for a conditional statement, for example:

```ruby
<%= render partial: "delete_#{@delete_logs.form.log_type}" %>
```

arfacamble replied:
aha ok cool. will do.
out of interest, I would expect that to be `render partial: ...`
a quick scan of docs and SO makes me think it doesn't massively matter the difference?

jamesrose replied:
Sorry, yes, typo!

jamesrose commented on app/views/logs/update_logs.html.erb:
```diff
@@ -4,43 +4,42 @@
+            scope: "row",
+            class: "govuk-!-text-align-right",
+          }) do %>
+            <%= govuk_link_to("Update now", send(log.form.unresolved_log_path, log)) %>
```
Using `send` here is superfluous:

```suggestion
            <%= govuk_link_to("Update now", log.form.unresolved_log_path(log)) %>
```

arfacamble replied:
I only changed the indentation in this file, I didn't read it in much detail, happy to change this.

jamesrose commented on app/models/forms/delete_logs_form.rb:
```diff
@@ -0,0 +1,40 @@
+      @search_term = attributes[:search_term]
+      @current_user = attributes[:current_user]
+      @logs = FilterService.filter_logs(all_logs, @search_term, attributes[:log_filters], nil, @current_user)
+      @selected_ids = attributes[:selected_ids] || @logs.map(&:id)
```
As far as I can see, `attributes[:selected_ids]` will always have values, or default to `[]`. Is that right?

If you fetch with a default then you can remove the defaulting outside of this class:

```suggestion
      @selected_ids = attributes.fetch(:selected_ids, [])
```

arfacamble replied:
There may be a better way to organise this:
- when the user initially navigates from an index page, there are no selected_ids in attributes and we wish to create an array with all ids in it (we can only create this array after calling the FilterService, which happens in the Form object initialize)
- if the user presses cancel on the confirmation page, he is returned with an array of selected_ids which are passed into the form object initialize
- if the user unselects all checkboxes and presses delete, then in the controller confirmation methods `params[:forms_delete_logs_form]` has no selected_ids key in it at all, so if it is not present, we set it to be an empty array, eg on line 25.
I can't immediately see a way to rejig what I've done to make it cleaner and account for these situations, but I'm happy to hear suggestions.

jamesrose commented on app/controllers/delete_logs_controller.rb:
```diff
@@ -0,0 +1,196 @@
+  end
+
+  def discard_sales_logs_for_organisation
+    logs = SalesLog.find(params.require(:ids))
```
If we know the organisation at this point, we should try and be defensive and scope our query under that organisation.

arfacamble replied:
I'm happy to do this. It seems sensible good practice but is it necessary more than that?
The ids are coming from the last page, so they are either correct, or the user is doing something devious, but the user can still only delete logs for which he is authorised to on the next line anyway.

jamesrose commented on app/controllers/delete_logs_controller.rb:
```diff
@@ -0,0 +1,196 @@
+    else
+      render "logs/delete_logs"
+    end
+  end
```
As you repeat the first block of this method, would you consider extracting the params out into a separate method?

```ruby
  def delete_sales_logs_for_organisation_confirmation
    @delete_logs_form = Forms::DeleteLogsForm.new(attributes_for_sales_logs(organisation: true))
    if @delete_logs_form.valid?
      render "logs/delete_logs_confirmation"
    else
      render "logs/delete_logs"
    end
  end
  
  private
  
  # ...
  
  def attributes_for_sales_logs(organisation: false)
     attributes = {
      current_user:,
      log_filters: @session_filters,
      log_type: :sales,
    }
    attributes.merge(organisation ? sales_logs_for_organisation_paths : sales_logs_paths)
  end 
```

arfacamble replied:
I hadn't fully looked through this suggested change, but I have refactored in a similar but different way.
I've retained the delete_logs_form private method, but made it able to be used by all controller methods wanting a delete logs form. I will push that along with a range of other changes soon

jamesrose commented on app/views/logs/delete_logs_confirmation.html.erb:
```diff
@@ -0,0 +1,25 @@
+<div class="govuk-button-group">
+  <%= govuk_button_to "Delete logs", @delete_logs_form.delete_path, method: "delete", params: { ids: @delete_logs_form.selected_ids } %>
+  <%= form_with url: @delete_logs_form.delete_path do |f| %>
+    <%= f.hidden_field :selected_ids, value: @delete_logs_form.selected_ids %>
```
You can pass these in as an array and it should just work:

```suggestion
    <% @delete_logs_form.selected_ids.each do |id|
      <%= f.hidden_field "selected_ids[]", value: id %>
    <% end %>
```

arfacamble replied:
I would expect both of these options to result in the same html. Is there a reason to want the more verbose one?

jamesrose commented on spec/requests/delete_logs_controller_spec.rb:
```diff
@@ -0,0 +1,946 @@
+      allow(FilterService).to receive(:filter_logs).and_return LettingsLog.all
+    end
+
+    it "calls the filter service with the filters in the session and the search term from the query params" do
```
You are testing the functionality of `FilterService` here which should be tested within its own spec.

arfacamble replied:
I'm interested in discussing this.
The FilterService is tested independently, my intention was to make sure that the method calls the FilterService with the correct arguments and then in the next test that the page displays whatever is returned.

arfacamble committed:
extract tables for lettings and sales to own partials

arfacamble committed:
refactor delete logs controller after tech review
improve the private method that builds the form object so that it has the flexibility to do so for all controller methods
ensure that the search term is passed to the delete logs controller when navigating through the organisations tab
ensure that noly logs for that organisation are displayed when navigating to delete logs through the organisations tab

arfacamble committed:
remove unnecessary untested arguments

arfacamble committed:
test new helper methods

arfacamble committed:
implement dirty fiddle to get the checkboxes smaller and also not misaligned

arfacamble committed:
ensure delete logs button is always visible on log lists when in the organisations tab

arfacamble committed:
minor linting corrections

arfacamble committed:
revert change, causing errors and outside the scope of this ticket

arfacamble committed:
simplify tests for whether delete logs button appears on index page

arfacamble committed:
replicate request specs from lettings for sales and organisations controllers

arfacamble committed:
minor refactor of lettings log feature spec setup, replicate happy path for sales

arfacamble committed:
minor refactors after rebasing onto Nat's work

arfacamble committed:
temp

arfacamble committed:
write tests for the delete logs form object

arfacamble committed:
lint: add new line at end of file

arfacamble committed:
respond to PO feedback
the log id in the delte logs table should be a link to the log
the delete logs button should be visible when the user is in a bulk upload journey
updated associated tests
